### PR TITLE
[Feature] Add diff seeds to diff ranks.

### PR DESCRIPTION
**Pull request by hhaAndroid**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
This demand comes from [mmselfsup](https://github.com/open-mmlab/mmselfsup/pull/228/). In the training of [MAE](https://arxiv.org/pdf/2111.06377.pd), random occlusion operations on the input image blocks are required. In order to ensure that each rank has different randomness, it is necessary to set different seeds for different ranks. 

There is no way to make different models run with different seeds in the previous implementation. So, it is necessary to set different seeds for different ranks at the beginning of the program and synchronize them in the sampler to ensure that the seeds are the same.

MMDetection does not have this demand for the time being, but in order to unify the various opennmmlab repo, we have also made corresponding modifications.

## BC-breaking (Optional)
No

### CODE REVIEW

File: mmdet/core/utils/dist_utils.py [hunk 1]
```diff
@@ -151,3 +152,28 @@ def all_reduce_dict(py_dict, op='sum', group=None, to_float=True):
     if isinstance(py_dict, OrderedDict):
         out_dict = OrderedDict(out_dict)
     return out_dict
+
+
+def shared_random_seed(seed=None, device='cuda'):
```
Comment (ZwwWayne) [inline mmdet/core/utils/dist_utils.py:157]: shared_random_seed -> sync_random_seed

File: tools/train.py [hunk 2]
```diff
@@ -168,7 +170,12 @@ def main():
     logger.info(f'Config:\n{cfg.pretty_text}')
 
     # set random seeds
-    seed = init_random_seed(args.seed)
+    # make sure each rank has a different
+    if args.seed is None:
+        seed = np.random.randint(2**31)
+    else:
+        seed = args.seed
+    seed = seed + dist.get_rank()
```
Comment (ZwwWayne) [inline tools/train.py:178]: make it an option and keep the default behavior.

Comment (RangiLyu) [inline tools/train.py:178]: I think we still need an option to enable all ranks to use the same seed.

File: tools/train.py [hunk 3]
```diff
@@ -52,6 +53,10 @@ def parse_args():
         help='id of gpu to use '
         '(only applicable to non-distributed training)')
     parser.add_argument('--seed', type=int, default=None, help='random seed')
+    parser.add_argument(
+        '--diff_seed',
```
Comment (RangiLyu) [inline tools/train.py:57]:
```suggestion
        '--diff-seed',
```

File: mmdet/datasets/samplers/infinite_sampler.py [hunk 4]
```diff
@@ -48,7 +50,7 @@ def __init__(self,
         self.world_size = world_size
         self.dataset = dataset
         self.batch_size = batch_size
-        self.seed = seed if seed is not None else 0
+        self.seed = sync_random_seed(seed) if seed is not None else 0
```
Comment (RangiLyu) [inline mmdet/datasets/samplers/infinite_sampler.py:53]:
```suggestion
        self.seed = sync_random_seed(seed)
```

Setting the seed defaults to 0 will reduce randomness after resume. Better to generate a random number as the default seed.

**Comment by jshilong** [comment]
Please check whether this pr would break down the initialization of the dataset in mmfewshot

File: tools/train.py [hunk 5]
```diff
@@ -169,6 +174,7 @@ def main():
 
     # set random seeds
     seed = init_random_seed(args.seed)
```
Comment (RangiLyu) [inline tools/train.py:176]: Can this be replaced with sync_random_seed?

Comment (jshilong) [reply tools/train.py:176]: May `init_random_seed` not be equal to`sync_random_seed`?

File: mmdet/datasets/samplers/infinite_sampler.py (hunk 4 above)
Comment (hhaAndroid) [reply mmdet/datasets/samplers/infinite_sampler.py:53]: Currently the default value of seed is 0, so if the user does not set it, it will be 0.

**Comment by hhaAndroid** [comment]
> Please check whether this pr would break down the initialization of the dataset in mmfewshot

should not, without  bc breaking.

File: tools/train.py (hunk 5 above)
Comment (hhaAndroid) [reply tools/train.py:176]: The logic is not the same, and bc breaking is not expected.

File: mmdet/datasets/samplers/infinite_sampler.py (hunk 4 above)
Comment (hhaAndroid) [reply mmdet/datasets/samplers/infinite_sampler.py:53]: At present, the program will definitely pass the seed

**Review by RangiLyu — APPROVED** [review]

**Pull request closed without merging**